### PR TITLE
PSR-454 PSR-674 Member Removal Issues fixed

### PR DESCRIPTION
**Pull request by B-Savage**

### CODE REVIEW

File: app/pages/nonsipp/memberdetails/MemberDetailsPage.scala [hunk 1]
```diff
@@ -29,15 +30,21 @@ import scala.util.Try
 case class MemberDetailsPage(srn: Srn, index: Max300) extends QuestionPage[NameDOB] {
 
   // won't work with a get all pages as Map as the arrayIndex must be a string
-  override def path: JsPath = Paths.personalDetails \ toString \ index.arrayIndex
+  override def path: JsPath = Paths.personalDetails \ toString \ index.arrayIndex.toString
 
   override def cleanup(value: Option[NameDOB], userAnswers: UserAnswers): Try[UserAnswers] =
-    value.fold(userAnswers.remove(DoesMemberHaveNinoPage(srn, index)))(_ => Try(userAnswers))
+    removePages(userAnswers, dependantPages(srn))
```
Comment (kberec) [inline app/pages/nonsipp/memberdetails/MemberDetailsPage.scala:36]: @B-Savage this removes dependent pages even if the details are getting updated.

Comment (kberec) [reply app/pages/nonsipp/memberdetails/MemberDetailsPage.scala:36]: Some commented code left.

File: app/pages/nonsipp/memberdetails/MemberDetailsPage.scala [hunk 2]
```diff
@@ -29,15 +30,44 @@ import scala.util.Try
...
+      case (Some(x), Some(y)) if (x == y) =>
+        // value not changed
+        Try(userAnswers)
+      case (Some(x), Some(y)) if (x != y) =>
+        // value changed
+        Try(userAnswers)
+      case (None, _) =>
+        //deletion
+        removePages(userAnswers, pages(srn))
+      case _ => Try(userAnswers)
```
Comment (kberec) [inline app/pages/nonsipp/memberdetails/MemberDetailsPage.scala:58]: This case covers the first three cases too, so the first three are unnecessary.

**Review by kberec — APPROVED** [review]
Discussed sorting issue since this was approved.

**Review by kberec — CHANGES_REQUESTED** [review]
Let's have a chat about issues with sorting

File: app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala [hunk 3]
```diff
@@ -63,7 +63,7 @@ class EmployerContributionsMemberListController @Inject()(
 
   def onPageLoad(srn: Srn, page: Int, mode: Mode): Action[AnyContent] = identifyAndRequireData(srn) {
     implicit request =>
-      val memberList = request.userAnswers.list(MembersDetailsPages(srn))
+      val memberList = request.userAnswers.map(MembersDetailsPages(srn)).values.toList
```
Comment (kberec) [inline app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala:66]: There must be a simplified version too, but IMO this should work:

```
val memberList = request.userAnswers.map(MembersDetailsPages(srn)).toList.sortBy(_._1.toInt).map(_._2)
```

**Review by kberec — CHANGES_REQUESTED** [review]

File: app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala (hunk 3 above)
Comment (kberec) [reply app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala:66]: Just to summarise our discussion:
Above is just an example, there must be some nicer solution. 
Also this doesn't fix the scenario where there is a gap in the map of members, the employer contribution is still bound to the wrong member.

File: app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala [hunk 4]
```diff
@@ -63,10 +62,14 @@ class EmployerContributionsMemberListController @Inject()(
 
   def onPageLoad(srn: Srn, page: Int, mode: Mode): Action[AnyContent] = identifyAndRequireData(srn) {
     implicit request =>
-      val memberList = request.userAnswers.map(MembersDetailsPages(srn)).values.toList
+      val memberMap = request.userAnswers.map(MembersDetailsPages(srn))
+      val maxIndex: Int = memberMap.keys.map(_.toInt).maxOption.getOrElse(-1)
```
Comment (cyeniceri) [inline app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala:66]: is this "-1" logic working properly when MembersDetailsPages returns empty ? I mean ((0 to -1))

File: app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala [hunk 5]
```diff
@@ -81,21 +84,24 @@ class EmployerContributionsMemberListController @Inject()(
 
   def onSubmit(srn: Srn, page: Int, mode: Mode): Action[AnyContent] = identifyAndRequireData(srn).async {
     implicit request =>
-      val memberList = request.userAnswers.membersDetails(srn).values.toList
+      val memberMap = request.userAnswers.map(MembersDetailsPages(srn))
+      val maxIndex: Int = memberMap.keys.map(_.toInt).maxOption.getOrElse(-1)
```
Comment (cyeniceri) [inline app/controllers/nonsipp/employercontributions/EmployerContributionsMemberListController.scala:88]: same here

File: test/controllers/nonsipp/memberdetails/SchemeMembersListControllerSpec.scala [hunk 6]
```diff
@@ -55,7 +55,7 @@ class SchemeMembersListControllerSpec extends ControllerBaseSpec {
             implicit request =>
               injected[ListView].apply(
                 form(injected[YesNoPageFormProvider], Manual),
-                viewModel(srn, 1, Manual, NormalMode, List((refineMV(1), memberDetails.fullName, Active)))
+                viewModel(srn, 1, Manual, NormalMode, List((refineMV(1), ("0", memberDetails.fullName), Active)))
```
Comment (kberec) [inline test/controllers/nonsipp/memberdetails/SchemeMembersListControllerSpec.scala:58]: Are these "0" values right here? If so, should it be a variable?

**Review by kberec — APPROVED** [review]

**Pull request merged**